File: assignment 5/python/programs/1.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class SinglyCircularLinkedList:
    def __init__(self):
        self.head = None

    # Insert a node at the end of the circular linked list
    def insert(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            self.head.next = self.head
        else:
            current = self.head
            while current.next != self.head:
                current = current.next
            current.next = new_node
            new_node.next = self.head

    # Display the circular linked list
    def display(self):
        if not self.head:
            print("Circular Linked List is empty.")
            return

        current = self.head
        while True:
            print(current.data, end=" -> ")
            current = current.next
            if current == self.head:
                break
        print()


if __name__ == "__main__":
    circular_linked_list = SinglyCircularLinkedList()

    # Insert elements into the circular linked list
    circular_linked_list.insert(1)
    circular_linked_list.insert(2)
    circular_linked_list.insert(3)
    circular_linked_list.insert(4)

    # Display the circular linked list
    print("Circular Linked List:")
    circular_linked_list.display()

File: assignment 5/c/programs/1.c
#include <stdio.h>
#include <stdlib.h>

struct Node
{
    int data;
    struct Node *next;
};

void addNode(struct Node **head_ref, int new_data)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));
    struct Node *last = *head_ref;

    new_node->data = new_data;
    new_node->next = *head_ref;

    if (*head_ref != NULL)
    {
        while (last->next != *head_ref)
            last = last->next;
        last->next = new_node;
    }
    else
        new_node->next = new_node;

    *head_ref = new_node;
}

void printList(struct Node *head)
{
    struct Node *temp = head;
    if (head != NULL)
    {
        do
        {
            printf("%d ", temp->data);
            temp = temp->next;
        } while (temp != head);
    }
}

int main()
{
    struct Node *head = NULL;

    addNode(&head, 12);
    addNode(&head, 56);
    addNode(&head, 2);
    addNode(&head, 11);

    printf("Contents of Circular Linked List\n ");
    printList(head);

    return 0;
}
